### PR TITLE
Fixes grad_dtype property being lost in fsdp

### LINKED ISSUE #170648: fsdp2 support grad_dtype
### Problem
The `grad_dtype` property of PyTorch tensors, which allows specifying gradient dtype for leaf tensors, is not preserved when applying `torch.distributed.fsdp.fully_shard`. This breaks precision control in distributed training.

### Alternatives

_No response_

### Additional context

_No response_

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #170648 

**Problem**
When `fully_shard()` is applied, it creates new `nn.Parameter` objects to hold the sharded data. The original parameter's `grad_dtype` property was not being copied to these new parameters, so any user-set `grad_dtype` was lost.

**Fix**

1. `_fsdp_param.py`: Before creating the sharded parameter, save the original `grad_dtype`. After creating it, restore that value.
2. `_fsdp_collectives.py`: Before assigning a gradient to a parameter, cast the gradient to match the parameter's grad_dtype if one is set.
3. Added test

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 1]
```diff
@@ -406,6 +408,8 @@ def _init_sharded_param(
             )
         self.sharded_param = nn.Parameter(self.to_sharded_dtensor(sharded_param))
         self.sharded_param.requires_grad_(param.requires_grad)
```
Comment (Skylion007) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: Requires param is in the nn.Parameter kwarg ctor, why not set it in the ctor?

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py [hunk 2]
```diff
@@ -388,6 +388,35 @@ def assert_fn(output: torch.Tensor):
...
+            self.assertEqual(param.grad_dtype, desired_grad_dtype)
+
+        torch.manual_seed(42 + self.rank + 1)
+        inp = torch.randn(2, 16, device=device_type, dtype=torch.bfloat16)
+        model(inp).sum().backward()
+
+        for param in model.parameters():
+            self.assertIsNotNone(param.grad)
+            self.assertEqual(param.grad.dtype, desired_grad_dtype)
+
```
Comment (ppwwyyxx) [inline test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:597]: 1. The parameters before sharding have grad_dtype set, `fully_shard()` shall preserve the property on sharded params.
2. In fsdp-allgather, the all-gathered full param shall preserve grad_dtype property. 

I think both are expected behavior. (1) is covered by the test, but could you also check (2) ? This would be an assertion in the layer's forward() to check that the full param is (i) no longer a DTensor and (i) has the correct `grad_dtype`.

Comment (arkadip-maitra) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:597]: Thanks for the suggestion! I looked into this.

The issue is that `grad_dtype`  causes autograd to compute gradients in that dtype. This makes reduce-scatter operate in `grad_dtype` (float32) instead of `param_dtype` (bfloat16), which breaks `test_compute_dtype` and `test_clamp_reduce_dtype`.

Modified the test slightly, so that it verifies `grad_dtype` is preserved on sharded params after fully_shard() and gradients have correct `grad_dtype` after backward.

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 3]
```diff
@@ -630,6 +630,12 @@ def foreach_reduce(
                     # Record an event on which to block the CPU thread to
                     # ensure that the D2H copy finishes before the optimizer
                     fsdp_param.grad_offload_event = post_reduce_stream.record_event()
+            param_grad_dtype = fsdp_param.sharded_param.grad_dtype
+            if (
+                param_grad_dtype is not None
+                and new_sharded_grad.dtype != param_grad_dtype
+            ):
+                new_sharded_grad = new_sharded_grad.to(param_grad_dtype)
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py]: maybe do the casting before line 617? it makes sure accurate grad dtype before offload_to_cpu

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 4]
```diff
@@ -404,8 +406,11 @@ def _init_sharded_param(
             raise AssertionError(
                 f"Expected contiguous tensor with {self.fsdp_placement=}"
             )
-        self.sharded_param = nn.Parameter(self.to_sharded_dtensor(sharded_param))
-        self.sharded_param.requires_grad_(param.requires_grad)
+        self.sharded_param = nn.Parameter(
+            self.to_sharded_dtensor(sharded_param), requires_grad=param.requires_grad
+        )
+        if should_restore_grad_dtype and self.sharded_param.requires_grad:
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: maybe this would be more compact? it removes the need for line 266-267
```
if param.is_leaf and param.requires_grad:
    self.sharded_param.grad_dtype = param.grad_dtype
```

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py [hunk 5]
```diff
@@ -388,6 +388,37 @@ def assert_fn(output: torch.Tensor):
             ):
                 ref_param_compute.detach().copy_(ref_param)
 
+    @skip_if_lt_x_gpu(2)
+    def test_grad_dtype_preserved(self):
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:462]: need to extend the unit tests
* input model: bf16 model with .grad_dtype fp32, fp32 model with .grad_dtype bf16
* fsdp2 config: no mixed preicison with reduce_dtype != .grad_dtype, enable mixed precision with reduce_dtype != .grad_dtype
* cover hsdp2, where reduce-scatter and all-reduce both happens
* cover gradient accumulation

**Review by weifengpy — CHANGES_REQUESTED** [review]
see comment above

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 3 above)
Comment (arkadip-maitra) [reply torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py]: yes

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py (hunk 4 above)
Comment (arkadip-maitra) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py]: yes

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py (hunk 5 above)
Comment (arkadip-maitra) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:462]: added

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py (hunk 2 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:597]: we expect same autograd behavior aross single-device and FSDP multi-device

from claude on what's different right now, this is related to @ppwwyyxx 's point on "the all-gathered full param shall preserve grad_dtype property"

```
Single-device grad_dtype semantics:                                                                           
   
  At engine.cpp:951-953, the cast happens in validate_outputs on the edge leading to AccumulateGrad. This runs  
  after backward computation, before gradient storage:                                                        

  backward ops (dtype determined by ops, NOT grad_dtype)
    → validate_outputs: cast to grad_dtype    ← HERE
      → AccumulateGrad: store/accumulate .grad

  grad_dtype does not affect backward compute dtype. It only affects the final .grad dtype. The author's claim
  that "grad_dtype causes autograd to compute gradients in that dtype" is incorrect.

  Now, does FSDP preserve this?

  The final .grad dtype — yes. The foreach_reduce cast at line 698 ensures sharded_param.grad.dtype ==          
  grad_dtype.
                                                                                                                
  But there's a subtle difference in the gradient accumulation path. Consider: float32 param,                   
  grad_dtype=bfloat16, no mixed precision, 3 microbatches:
                                                                                                                
  Single-device:                                                                                                
  1. Backward → float32 grad → AccumulateGrad casts to bfloat16 → stores .grad
  2. Backward → float32 grad → cast to bfloat16 → += in bfloat16                                                
  3. Same. Final .grad is bfloat16, accumulated in bfloat16.                                                  
                                                                                                                
  FSDP (HEAD):                                                                                                  
  1. Backward → float32 grad → AccumulateGrad stores float32 (no cast, _unsharded_param.grad_dtype is default   
  float32)                                                                                                      
  2. accumulate_unsharded_grad_if_needed → += in float32                                                      
  3. Final reduce-scatter → cast to bfloat16 in foreach_reduce.                                                 
                                                                                                                
  Same final dtype, but accumulation happened in float32 instead of bfloat16. Numerically different.            
                                                                                                                
  The fix ppwwyyxx suggested — setting grad_dtype on _unsharded_param — would make AccumulateGrad cast before   
  FSDP sees the gradient, matching single-device accumulation semantics. It wouldn't change backward compute    
  dtype (that's determined by the ops, not grad_dtype). But it would change the dtype of the gradient FSDP feeds
   into reduce-scatter, which is why existing tests break — the reduce-scatter dtype logic assumes the gradient
  is in param_dtype, not grad_dtype.
```

**Review by weifengpy — COMMENTED** [review]
see comment on .grad_dtype on unsharded parameters above

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py (hunk 2 above)
Comment (arkadip-maitra) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:597]: Hey wei, thanks for the comment. Updated the code.

Propagated `grad_dtype to _unsharded_param`, so AccumulateGrad casts gradients before FSDP sees them.
Also `casted grad_dtype to reduce_dtype` before copy_in and `reduce_dtype to grad_dtype` on the sharded output.
Also added test to verify numerical parity with single device behavior.

**Review by weifengpy — COMMENTED** [review]
see claude code review above

**Comment by ppwwyyxx** [comment]
@weifengpy could you review again?

**Comment by junjzhang** [comment]
any updates?

**Comment by arkadip-maitra** [comment]
> any updates?

i need to resolve the merge conflicts and then get it reviewed again.

**Comment by weifengpy** [comment]
rebase? I will take a look again today. if things are not still not working, i might publish marking @arkadip-maitra as the co-author

**Comment by cyyever** [comment]
help rebase？

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py [hunk 6]
```diff
@@ -458,6 +458,240 @@ def _reduce_1d_partial_grads(
             if param.grad is not None:
                 param.grad.div_(group.size())
 
+    @skipIfRocmVersionLessThan((7, 0))
+    @skip_if_lt_x_gpu(2)
+    @requires_nccl_version((2, 10), "Need NCCL 2.10+ for bf16 collectives")
```
Comment (cyyever) [inline test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py]: these constraints may be stale.

Comment (arkadip-maitra) [reply test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py]: true, updated, thanks for the comment!!

**Review by cyyever — APPROVED** [review]
lgtm

**Review by weifengpy — CHANGES_REQUESTED** [review]
i am confirming on 2 things from claude
* grad accumulation silently downcasts to `reduce_dtype`, defeating the feature — `_fsdp_param.py:1093`
* hard crash on float8 grads


i do think .grad_dtype is very important and there has been several rounds of reviews. i am speeding this up by doing a co-author PR with arkadip now

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 7]
```diff
@@ -600,7 +600,20 @@ def foreach_reduce(
         device=device,
     )
 
-    foreach_reduce_scatter_copy_in(unsharded_grads, reduce_scatter_input, world_size)
+    # _chunk_cat handles widening casts (e.g. bf16 -> fp32) natively but
+    # rejects narrowing casts. Use a temp buffer only for narrowing.
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:604]: is the _chunk_cat claim correct? _chunk_cat fp32 → bf16 didn't raise error

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 8]
```diff
@@ -600,7 +600,20 @@ def foreach_reduce(
...
+    if (
+        grad_dtype != reduce_dtype
+        and torch.promote_types(grad_dtype, reduce_dtype) != reduce_dtype
+    ):
+        grad_buffer = torch.empty_like(reduce_scatter_input, dtype=grad_dtype)
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:609]: generally we are very sensitive in torch.empty_like and any tensor creations during forward/backward. we intentionally allocate buffers beforehand at init time

bigger questions is: removing this branch and just do foreach_reduce_scatter_copy_in is bit-wise equal: https://github.com/pytorch/pytorch/pull/194434/changes#diff-6e7c312d014961bab53c1600f2ed2b5762df2afd00f289271995165ca7919ee5L617

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 9]
```diff
@@ -693,7 +706,14 @@ def foreach_reduce(
         # AR to finish. The reduce-dtype buffer is held across layers by
         # FSDPParamGroup._all_reduce_state (captured above) to prevent
         # this. See PR #140044, regression test PR #180900.
-        reduce_output = _to_dtype_if_needed(reduce_output, orig_dtype)
+        # Cast directly to grad_dtype if set, skipping orig_dtype to
+        # avoid a lossy round-trip (e.g. fp32 reduce -> bf16 orig -> fp32).
+        target_dtype = orig_dtype
+        if fsdp_params:
+            param_grad_dtype = fsdp_params[0].sharded_param.grad_dtype
+            if param_grad_dtype != fsdp_params[0].sharded_param.dtype:
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:714]: should handle `grad_dtype = None`, otherwise grad_dtype None makes this true, target_dtype = None, and _to_dtype_if_needed no-ops

**Comment by weifengpy** [comment]
drafting co-authored PR here: https://github.com/pytorch/pytorch/pull/194434
